lightning/src/ln/peers/handler.rs

```diff
@@ -650,6 +650,85 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, TransportImpl
+			log_info!(self.logger, "Peer global features required unknown version bits");
+			return Err(PeerHandleError { no_connection_possible: true }.into());
+		}
+		if init_message.features.requires_unknown_bits() {
+			log_info!(self.logger, "Peer local features required unknown version bits");
+			return Err(PeerHandleError { no_connection_possible: true }.into());
+		}
+
+		log_info!(
+			self.logger, "Received peer Init message: data_loss_protect: {}, initial_routing_sync: {}, upfront_shutdown_script: {}, static_remote_key: {}, unknown flags (local and global): {}",
+			if init_message.features.supports_data_loss_protect() { "supported" } else { "not supported"},
+			if init_message.features.initial_routing_sync() { "requested" } else { "not requested" },
+			if init_message.features.supports_upfront_shutdown_script() { "supported" } else { "not supported"},
+			if init_message.features.supports_static_remote_key() { "supported" } else { "not supported"},
+			if init_message.features.supports_unknown_bits() { "present" } else { "none" }
+		);
+
+		let sync_status = if init_message.features.initial_routing_sync() {
+			InitSyncTracker::ChannelsSyncing(0)
+		} else {
+			InitSyncTracker::NoSyncRequested
+		};
+
+		if !init_message.features.supports_static_remote_key() {
+			log_debug!(self.logger, "Peer {} does not support static remote key, disconnecting with no_connection_possible", log_pubkey!(their_node_id));
+			return Err(PeerHandleError { no_connection_possible: true }.into());
+		}
+
+		Ok(PostInitState::new(sync_status, init_message.features.clone()))
+	}
+
+	// Add an Init message to the outbound queue
+	fn queue_init_message(&self, peers_needing_send: &mut HashSet<Descriptor>, message_queuer: &mut impl MessageQueuer, pending_outbound_buffer: &mut impl PayloadQueuer, descriptor: &Descriptor, should_request_full_sync: bool) {
+		let mut features = InitFeatures::known();
+		if !should_request_full_sync {
+			features.clear_initial_routing_sync();
+		}
+
+		let resp = msgs::Init { features };
+		self.enqueue_message(peers_needing_send, message_queuer, pending_outbound_buffer, descriptor, &resp);
+	}
+
+	// Process an incoming Init message and set Peer and PeerManager state accordingly
+	fn process_init_message(&self, message: Message, descriptor: &Descriptor, peers_needing_send: &mut HashSet<Descriptor>, node_id_to_descriptor: &mut HashMap<PublicKey, Descriptor>, pending_outbound_buffer: &mut impl PayloadQueuer, outbound: bool, transport: &mut impl ITransport, post_init_state: &mut Option<PostInitState>) -> Result<(), PeerHandleError> {
+		let their_node_id = transport.get_their_node_id();
+
+		match message {
+			Message::Init(ref init_message) => {
+				log_trace!(self.logger, "Received Init message from {}", log_pubkey!(&their_node_id));
+				if node_id_to_descriptor.contains_key(&their_node_id) {
+					log_trace!(self.logger, "Got second connection with {}, closing", log_pubkey!(&their_node_id));
```

Comment:
Ya, that seems reasonable. I added a review patch `review: Catch duplicate connection earlier` to catch it at the completion of the NOISE handshake instead of the INIT message processing.